src/json/json.cc

```
@@ -262,6 +262,95 @@ STATIC JsonUtilCode parseSetCmdArgs(ValkeyModuleString **argv, const int argc, S
+        return JSONUTIL_WRONG_NUM_ARGS;
+    }
+
+    *num_keys = (argc - 1) / 3;
+    // Allocate memory for args_list 
+    *args_list = reinterpret_cast<MSetCmdArgs *>(ValkeyModule_Alloc((*num_keys) * sizeof(MSetCmdArgs)));
+    memset(*args_list, 0, (*num_keys) * sizeof(MSetCmdArgs));
+
+    if (!(*args_list)) {
+        return JSONUTIL_ALLOCATION_FAILURE;
+    }
+
+    JsonUtilCode rc;
+    // Parse and validate arguments for each key
+    size_t i;
+    for (i = 0; i < *num_keys; ++i) {
+        MSetCmdArgs &current_arg = (*args_list)[i];
+
+        current_arg.key_str = argv[i * 3 + 1];
+        current_arg.key = static_cast<ValkeyModuleKey*>(
+            ValkeyModule_OpenKey(ctx, current_arg.key_str, VALKEYMODULE_READ | VALKEYMODULE_WRITE));
+
+        // Handle key allocation failure
+        if (!current_arg.key) {
+            rc = JSONUTIL_KEY_OPEN_ERROR;
+            return rc;
+        }
+
+        current_arg.path = ValkeyModule_StringPtrLen(argv[i * 3 + 2], nullptr);
+        current_arg.json = ValkeyModule_StringPtrLen(argv[i * 3 + 3], &current_arg.json_len);
+
+        // Validate key type
+        int type = ValkeyModule_KeyType(current_arg.key);
+        if (type != VALKEYMODULE_KEYTYPE_EMPTY &&
+            ValkeyModule_ModuleTypeGetType(current_arg.key) != DocumentType) {
+            rc = JSONUTIL_NOT_A_DOCUMENT_KEY;
+            return rc;
+        }
+
+        // Root path validation
+        bool is_new_valkey_key = (type == VALKEYMODULE_KEYTYPE_EMPTY);
+        current_arg.is_root_path = jsonutil_is_root_path(current_arg.path);
+        if (is_new_valkey_key && !current_arg.is_root_path) {
+            rc = JSONUTIL_COMMAND_SYNTAX_ERROR;
+            return rc;
+        }
+
+        // Validate JSON structure
+        if (current_arg.is_root_path) {
+            JDocument *doc = nullptr;
+            rc = dom_parse(ctx, current_arg.json, current_arg.json_len, &doc);
+            ValkeyModule_Assert(rc == JSONUTIL_SUCCESS);
```

Cannot use assert here because the json input string could be invalid. Need to return the error code. Change the line to:
```
if (rc != JSONUTIL_SUCCESS) {
  if (doc) dom_free_doc(doc);
   return rc;
}
```